src/futures/digital.rs

```diff
@@ -0,0 +1,57 @@
+//! Asynchronous digital I/O
+
+use core::future::Future;
+
+use crate::blocking::digital::InputPin;
+
+/// Asynchronously wait for a pin to become high or low.
+///
+/// # Examples
+/// ```rust
+/// # use embedded_hal::futures::digital::AsyncInputPin;
+/// /// Asynchronously wait until the `ready_pin` becomes high.
+/// async fn wait_until_ready<P>(ready_pin: &P)
+/// where
+///     P: AsyncInputPin,
+/// {
+///     ready_pin
+///         .until_high()
+///         .await
+///         .expect("failed to await input pin")
+/// }
+/// ```
+///
+/// ```rust,ignore
+/// # use embedded_hal::futures::digital::AsyncInputPin;
+/// # use embedded_hal::futures::delay::Delay;
+/// use core::time::Duration;
+///
+/// /// Wait until the `ready_pin` is high or timeout after 1 millisecond.
+/// /// Returns true is the pin became high or false if it timed-out.
+/// async fn wait_until_ready_or_timeout<P, D>(ready_pin: &P, delay: &mut D) -> bool
+/// where
+///     P: AsyncInputPin,
+///     D: Delay,
+/// {
+///     futures::select_biased! {
+///         x => ready_pin.until_high() => {
+///             x.expect("failed to await input pin");
+///             true
+///         },
+///         _ => delay.delay(Duration::from_millis(1)) => false, // ignore the error
+///     }
+/// }
+/// ```
+pub trait AsyncInputPin: InputPin {
```

Comment:
- What's the benefit of having the `InputPin` bound? Users can bound on both if they want both, it seems a bit unnecessary restriction to me.
- Name shouldn't have "async", all traits in the futures module are async
- Name being "InputPin" suggests this is the async equivalent of InputPin, namely having is_high/is_low functions but async. That's not what the trait does. I like the naming in `embassy_traits` much more: `WaitForHigh`, `WaitForLow`. It makes it clear what it does, and it makes method names to be verbs like all the other traits.

Reply:
These are all good points. I'm not sure it should be two different traits (e.g. `WaitForHigh`, `WaitForLow`) though. Maybe `WaitForPin`?

Reply:
No idea! there might be some cursed hardware out there that assumes external interrupts are active-low or something, so I made it 2 separate traits just in case in embassy. All the other traits are one trait per function too.

Reply:
One other thing: if this trait doesn't have an `InputPin` bound, then it _could_ have a different error type, which would result in people using the `P: InputPin + WaitFor<Error = <P as InputPin>::Error>` (if that even works) bound. I'm not a fan of there being an implicit agreement that these have the same error type.

Reply:
True, but this is already the case with SPI Read/Write/Transfer/TransferInplace. I don't think this is a problem...

Reply:
> there might be some cursed hardware out there that assumes external interrupts are active-low or something, so I made it 2 separate traits just in case in embassy.

Seems reasonable. Will do two traits then.

Reply:
> One other thing: if this trait doesn't have an `InputPin` bound, then it _could_ have a different error type, which would result in people using the `P: InputPin + WaitFor<Error = <P as InputPin>::Error>` (if that even works) bound. I'm not a fan of there being an implicit agreement that these have the same error type.

Maybe this trait shouldn't have an error type?